File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;

        // Edge case, if the root doesn't exist, we return empty vector.
        if (!root) {
            return result;
        }

        // Else we do a BFS with a queue data structure.
        // Since it's a tree, it's not going to contain cycles, so we can avoid the visited
        // information.
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            // process each level by level.
            vector<int> nodes_at_level;
            const int q_size = q.size();
            // pop these many elements.
            for (int i = 0; i < q_size; i++) {
                TreeNode* front = q.front();
                q.pop();
                nodes_at_level.push_back(front -> val);
                if (front -> left) q.push(front -> left);
                if (front -> right) q.push(front -> right);
            }
            result.push_back(nodes_at_level);
        }

        return result;
    }
};

File: Data Structures & Algorithms/same-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isEqualSubtree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p and q:
            if p.val != q.val:
                return False
            return self.isEqualSubtree(p.left, q.left) and self.isEqualSubtree(p.right, q.right)
        if (p and not q) or (q and not p):
            return False
        return True
        
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        return self.isEqualSubtree(p, q)
        